src/Microsoft.DocAsCode.Build.Engine/PostProcessors/PostProcessorsHandlerWithIncremental.cs

```diff
@@ -42,8 +46,152 @@ public void Handle(List<PostProcessor> postProcessors, Manifest manifest, string
+                CheckNoIncrementalItems(manifest, "After processing");
+            }
+            TraceIntermediateInfo(outputFolder, increItems, noneIncreItems);
+            PostHandle(manifest, increItems);
+        }
+
+        #region Handle related
+
+        private void PreHandle(Manifest manifest, string outputFolder, List<ManifestItem> increItems, List<ManifestItem> noneIncreItems)
+        {
+            if (_increContext.CanIncremental)
+            {
+                CopyToOutput(increItems, outputFolder);
+
+                // Copy none incremental items to post processors
+                manifest.Files = noneIncreItems.ToList();
+            }
+            else
+            {
+                // If cannot incremental post process, set all incremental flags to false
+                foreach (var item in manifest.Files)
+                {
+                    item.IsIncremental = false;
+                }
+                noneIncreItems.AddRange(increItems);
+                increItems.Clear();
+            }
+        }
+
+        private void PostHandle(Manifest manifest, List<ManifestItem> increItems)
+        {
+            if (_increContext.CanIncremental)
+            {
+                // Add back incremental items
+                manifest.Files.AddRange(increItems);
+            }
+        }
+
+        #endregion
+
+        #region Trace intermediate info
+
+        private void TraceIntermediateInfo(string outputFolder, List<ManifestItem> increItems, List<ManifestItem> noneIncreItems)
+        {
+            if (_increContext.ShouldTraceIncrementalInfo)
+            {
+                TraceIncremental(increItems);
+                TraceNoneIncremental(outputFolder, noneIncreItems);
+            }
+        }
+
+        private void TraceIncremental(List<ManifestItem> increItems)
+        {
+            foreach (var outputRelPath in GetOutputRelativePaths(increItems))
+            {
+                string lastCachedRelPath;
+                if (_increContext.LastInfo == null)
+                {
+                    throw new BuildCacheException("Last incremental post processor info should not be null.");
+                }
+                if (!_increContext.LastInfo.PostProcessOutputs.TryGetValue(outputRelPath, out lastCachedRelPath))
+                {
+                    throw new BuildCacheException($"Last incremental post processor outputs should contain {outputRelPath}.");
+                }
+
+                IncrementalUtility.RetryIO(() =>
+                {
+                    var lastCachedFile = Path.Combine(_increContext.LastBaseDir, lastCachedRelPath);
+                    var currentCachedFileName = IncrementalUtility.GetRandomEntry(_increContext.CurrentBaseDir);
+
+                    // Copy last cached file to current cached file
+                    File.Copy(lastCachedFile, Path.Combine(_increContext.CurrentBaseDir, currentCachedFileName));
```

Comment:
in last step of Build in DocumentBuilder, refer to [https://github.com/dotnet/docfx/blob/dev/src/Microsoft.DocAsCode.Build.Engine/DocumentBuilder.cs#L158](https://github.com/dotnet/docfx/blob/dev/src/Microsoft.DocAsCode.Build.Engine/DocumentBuilder.cs#L158)

---
In reply to: [95303401](https://github.com/dotnet/docfx/pull/1136#discussion_r95303401) [](ancestors = 95303401)